### PR TITLE
A foundation for contrastive self-supervised learning

### PR DESCRIPTION
**Pull request by CSJDeveloper**
**Contrastive Self-Supervised Learning (SSL)** has attracted great attention for its fantastic data efficiency and generalization ability in both computer vision and natural language processing. It learns the representation through a large number of unlabeled data and artificially defined self-supervision signals (i.e., regarding the augmented views of a data sample as positive samples). The model is updated by encouraging the embedding of positive samples close to each other. 

To support Plato's Contrastive Self-Supervised Learning (SSL), we implemented a foundation for contrastive Self-supervised learning.

## Description

In the commonly contrastive self-supervised learning (SSL), which relies on the centrally available data, the learning process is split into two stages because the core idea of SSL is to learn a generalizable representation from the data to behave as the input for the downstream tasks. Therefore, in the first stage, the SSL method performs on the trainset without using the supervised information, such as labels or annotations. This introduces a requirement for data augmentation methods (see our code in `plato/datasource/augmentations`) to prepare the paired samples, such as paired positive or triple (two positives and one negative) samples; thereby, the following contrastive learning can be conducted on these prepared samples to learn representation by solely relying on the samples metric instead of the supervised information. Once the training of the SSL method finishes, its performance should be measured based on the specific objective function of the downstream tasks. During this second stage, the responsibility of the SSL method is solely to extract representation from the tasks' train dataset while freezing the parameters. Then, the user is required to design a task-specific model capable of completing the tasks' objective based on the learned representation from the SSL method. Therefore, the obtained loss from the objective function is utilized to optimize the parameters of the designed task-specific model. Intuitively, as the SSL method is trained to extract the most representative features, the task-specific model relying on these features can be significantly lightweight, such as a one-layer Multilayer perception (MLP), to obtain great performance.
In summary, these two stages are:

1. Train the contrastive SSL methods solely on the samples without using any supervised information, such as labels or annotations. 
2. Evaluate the contrastive SSL method by applying the trained SSL model to the downstream tasks to complete the task-oriented objective function. This is generally referred to as the linear evaluation. Specifically, the learned representation from the SSL model behaves as the input for the lightweight task-specific model. The optimization is performed only in the task-specific model while maintaining the SSL model frozen. 

In the upper mentioned content, we only present how to train and measure the performance of the contrastive SSL method without touching on how to conduct the test during the two stages. After reading multiple codes from other trustworthy researchers who present detailed experiments and settings on different datasets, I categorize the ways to perform tests in these two stages into two prototypes. For simplicity, we use the classification task to describe without loss of generality.

1. Test the contrastive SSL method in stage one. This test is denoted as the learning monitor as it tracks the quality of learned representation on the current dataset. Therefore, such a monitor is generally designed to be the **k-nearest neighbors (KNN)** method. Specifically, the encoder of contrastive SSL extracts representation for all samples of the training dataset without using the data augmentation transform performed in the contrastive SSL method's training phase. These representations are utilized for training the KNN. Then, the KNN classifies the test dataset to obtain the label predictions to compute the accuracy based on the test dataset's ground-truth labels.
2. Test the contrastive SSL method in stage two. This test phase heavily relies on the objective of downstream tasks. In the image classification, the trained task-specific model is tested by performing on the test dataset to obtain the accuracy.

I observed one phenomenon from the existing source code and the corresponding works: they generally build two isolated files for these two stages, respectively. The users generally need to run the file for stage one to train the SSL model before switching to the file for stage two. However, under the federated learning of the Plato platform, it is pretty inconvenient for the user to write another piece of code to conduct the second stage. A more reasonable way is to perform stage two during stage one. One direct advantage is that the users only need to run the program once to obtain sufficient experiments.

Based on these insights and the noticed problems, I implemented the federated contrastive self-supervised learning (FedSSL) on Plato, following the designs below: 

For the `ssl_simple` client, it performs the general SSL train process within the `async def train(self)` function. Then, the current code implements two tests:
 
1.  A. `self.trainer.test` conducts the monitor test mentioned above. 
2. B.  `self.trainer.eval_test` conducts the whole stage two, including the task-specific training and the test. 
    
    Both of these two functions produce accuracy as the metric. 
    Then, in the configuration file, the user is expected to set  `do_test: True` to allow accuracy computation. Then, the  `test_interval: 3` can be set to determine the frequency of performing test A. In this example, the client performs test A every three communication rounds. Finally, the  `eval_test_interval: 10` can be set to determine the frequency of performing test B. In this example, the client performs test B every ten communication rounds.
    
The client holds three types of datasets and three types of data augmentations to perform these tests correctly. 
 1.  `self.trainset`. This is the general trainset for the SSL method. It occupies the data augmentation transform to prepare the paired samples for contrastive learning. The support dataset is the training dataset of this client. 
 2.  ` self.testset`. This is the general 'testset' for the monitoring process of the SSL method. It does not perform any data augmentation transform but only utilizes a simple transform for general classification tasks. The support dataset is the test dataset of this client. 
 3. `self.monitor_trainset`. This utilizes the same transform function as the self.testset, but the supporting dataset is the training dataset of this client. 
4. `self.eval_trainset`. This utilizes the transform function applied by the general image classification task. The support dataset is the training dataset of this client. 

There are two types of models in one client. 
1. The `self.model`. This is the SSL model used to perform stage one of self-supervised learning. This model is utilized as the global model to be sent to the server and shared among clients.
2. The `self.personalized_model`. This is the personalized model held by the client without being exposed to others. This can be regarded as the upper mentioned task-specific model for stage two of self-supervised learning. The `self.personalized_model` is only optimized during the `self.trainer.eval_test`. 

For the `contrastive_ssl`, it mainly contains four important functions. 

1.  `def train_loop()` performs the training process of SSL.
2.  `def test_process()` performs the monitor process of SSL.
4.  `def eval_test_process()` performs the whole linear evaluation process, i.e., the stage two.
5. `def eval_test()` performs the `def eval_test_process()`.

Other important model/data saving mechanisms.
We did not make any changes to the original Plato model and result storage. But, we added more code to support the personalized model/accuracy saving. There will be a dir named by 'Client_{id}' to save all experiments for this client.

Another important functions that we added to the Plato is the `get_dynamic_optimizer` and `get_dynamic_lr_schedule` in `optimizers.py`.
We hope to maintain multiple optimizers and schedulers in one client. It requires a more flexible way to support initial optimizers and the schedulers. However, limited by time, we did not want to redesign Plato's `get_optimizer` and conducted more tests on the previous code/methods. Thus, we just simply implemented our own methods. 

However, here is one concern of our code:
We did not add the implemented `ssl_simple` and `contrastive_ssl` to the corresponding registries of Plato. The main reason is that more rigorous should be performed before adding a new client and trainer to Plato. Obviously, our implementation does not satisfy this requirement. Then, the same reason for the document.

**The main principle of our implementation is to reuse Plato's existing code and implementation as much as possible.**

## How has this been tested?

First of all, we implemented a famous method, called SimCLR based on the above foundations of contrastive SSL. Then, we provided three experiments for code measurements.

A. `python examples/simclr/simclr_mnist_encoder.py -c examples/simclr/configs/simclr_MNIST_central.yml`

B. `python examples/simclr/simclr.py -c examples/simclr/configs/simclr_MNIST.yml`

C. `python examples/simclr/simclr.py -c examples/simclr/configs/simclr_CIFAR10_resnet18.yml` 

In A, we strictly followed the trusted implementation of a central SimCLR on the MNIST dataset. The same model structure, data augmentation, and optimization were utilized in our implementation. Then, we performed the test on 1 client to simulate the central learning (see the config file). This test is to ensure that our implementation of SimCLR is correct.

In B, we utilized the general way of contrastive SSL to design the encoder, i.e., an existing network without the final fully-connected layers. The test was performed on the MNSIT dataset under the LeNet5 network (see `encoders_register.py` for details). 

In C, We conducted a test of SSL on the CIFAR10 dataset with the ResNet18 without final fully-connected layers (see `encoders_register.py` for details) as the backbone. 

For the personalized model of these three tests, I utilized the purely one-layer Multilayer perception (MLP) network. Please access `general_mlps_register.py` for details. In this `general_mlps_register`, we implement the most flexible and common way to define the MLP networks. The main reason is that MLP networks are the necessary component for any SSL methods. However, the MLP networks of different SSL methods follow the same architecture but contain different combinations of modules, such as batch normalization and activations. Motivated by this phenomenon, we implemented this function to support the easiest way to define different MLPs for different methods. By the way, in this implementation, we also discussed the topic of "general order of components in one MLP layer". This is a very interesting topic that was frequently asked about by researchers. As mentioned in the comment, we finally chose to follow Schema A. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by baochunli** [comment]
Module `thop` is not a part of the requirements in Plato. Even if it were to be added to `requirements.txt`, it requires another package `onnx`, which is not a part of the requirements in Plato. Please remove the dependencies on `thop` as much as possible.

Custom models are passed into Plato as a class, not an instance, in PR #201. Also see `docs/Design.md`.

**Comment by CSJDeveloper** [comment]
> Module `thop` is not a part of the requirements in Plato. Even if it were to be added to `requirements.txt`, it requires another package `onnx`, which is not a part of the requirements in Plato. Please remove the dependencies on `thop` as much as possible.
> 
> A custom model is passed into Plato as a class, not an instance, in PR #201. Also see `docs/Design.md`.

I will remove `thop` as it was simply used as the tool to present the size and the number of parameters in the personalized model. Then, I will merge the main to my branch and modify the code.

**Comment by CSJDeveloper** [comment]
Closed this pull request as the new version has been created and will be uploaded
